### PR TITLE
Refactor GRoI to support concat mode

### PR DESCRIPTION
**Pull request by xvjiarui**

### CODE REVIEW

File: tests/test_roi_extractor.py [hunk 1]
```diff
@@ -39,3 +41,78 @@ def test_groie():
...
+    feats = (
+        torch.rand((1, 256, 200, 336)),
+        torch.rand((1, 256, 100, 168)),
+        torch.rand((1, 256, 50, 84)),
+        torch.rand((1, 256, 25, 42)),
+    )
+
+    rois = torch.tensor([[0.0000, 587.8285, 52.1405, 886.2484, 341.5644]])
+
+    with pytest.raises(AssertionError):
```
Comment (ZwwWayne) [inline tests/test_roi_extractor.py:117]: Add comments on what error you are testing.

File: tests/test_roi_extractor.py [hunk 2]
```diff
@@ -39,3 +41,78 @@ def test_groie():
...
+        torch.rand((1, 256, 25, 42)),
+    )
+
+    rois = torch.tensor([[0.0000, 587.8285, 52.1405, 886.2484, 341.5644]])
+
+    res = groie(feats, rois)
+    assert res.shape == torch.Size([1, 256, 7, 7])
+
+    # test w.o. pre/post concat
+    cfg = mmcv.Config(
```
Comment (ZwwWayne) [inline tests/test_roi_extractor.py:67]: We may not need `mmcv.Config`? Directly pass a dict should work.

File: tests/test_roi_extractor.py [hunk 3]
```diff
@@ -39,3 +41,78 @@ def test_groie():
 
     res = groie(feats, rois)
     assert res.shape == torch.Size([1, 256, 7, 7])
+
+    # test w.o. pre/post
+    cfg = mmcv.Config(
```
Comment (ZwwWayne) [inline tests/test_roi_extractor.py:46]: Same question as that on line 67.

File: mmdet/models/roi_heads/roi_extractors/generic_roi_extractor.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+    https://arxiv.org/abs/2004.13665
+
+    Args:
+        aggregate_type (str): The method to aggregate multiple feature maps.
+            Options are 'sum', 'concat'. Default: 'sum'.
+        pre_cfg (dict|None): Specify pre-processing modules. Default: None.
+        post_cfg (dict|None): Specify post-processing modules. Default: None.
+        kwargs (keyword arguments): Arguments that are the same
+            as :class:`SingleRoIExtractor`.
```
Comment (hellock) [inline mmdet/models/roi_heads/roi_extractors/generic_roi_extractor.py:26]: The docstring is outdated.

File: mmdet/models/roi_heads/roi_extractors/generic_roi_extractor.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    """
+
+    def __init__(self,
+                 aggregate_type='sum',
```
Comment (hellock) [inline mmdet/models/roi_heads/roi_extractors/generic_roi_extractor.py:30]: `aggregation='sum'`

File: mmdet/models/roi_heads/roi_extractors/generic_roi_extractor.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+from mmdet.models.builder import ROI_EXTRACTORS
+from mmdet.ops.plugin import build_plugin_layer
+from .base_roi_extractor import BaseRoIExtractor
+
+
+@ROI_EXTRACTORS.register_module()
+class GenericRoIExtractor(BaseRoIExtractor):
+    """Extract RoI features from all summed feature maps levels.
+
+    https://arxiv.org/abs/2004.13665
```
Comment (hellock) [inline mmdet/models/roi_heads/roi_extractors/generic_roi_extractor.py:18]: Introduce the title of the paper.

File: mmdet/models/roi_heads/roi_extractors/single_level_roi_extractor.py [hunk 7]
```diff
@@ -1,16 +1,15 @@
...
 
 
 @ROI_EXTRACTORS.register_module()
-class SingleRoIExtractor(nn.Module):
+class SingleRoIExtractor(BaseRoIExtractor):
     """Extract RoI features from a single level feature map.
 
-    If there are mulitple input feature levels, each RoI is mapped to a level
+    If there are multiple input feature levels, each RoI is mapped to a level
     according to its scale.
```
Comment (hellock) [inline mmdet/models/roi_heads/roi_extractors/single_level_roi_extractor.py:13]: Also add the paper title and link.

File: mmdet/models/roi_heads/roi_extractors/single_level_roi_extractor.py [hunk 8]
```diff
@@ -20,34 +19,9 @@ class SingleRoIExtractor(nn.Module):
...
-    def __init__(self,
-                 roi_layer,
-                 out_channels,
-                 featmap_strides,
-                 finest_scale=56):
-        super(SingleRoIExtractor, self).__init__()
-        self.roi_layers = self.build_roi_layers(roi_layer, featmap_strides)
-        self.out_channels = out_channels
-        self.featmap_strides = featmap_strides
+    def __init__(self, finest_scale=56, **kwargs):
```
Comment (hellock) [inline mmdet/models/roi_heads/roi_extractors/single_level_roi_extractor.py:22]: Keep the argument order consistent with the previous version.

File: mmdet/models/roi_heads/roi_extractors/generic_roi_extractor.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+        # mark the starting channels for concat mode
+        start_channels = 0
+        for i in range(num_levels):
+            roi_feats_t = self.roi_layers[i](feats[i], rois)
+            end_channels = start_channels + roi_feats_t.size(1)
+            if self.with_pre:
+                # apply pre-processing to a RoI extracted from each layer
+                roi_feats_t = self.pre_module(roi_feats_t)
+                roi_feats_t = self.relu(roi_feats_t)
```
Comment (hellock) [inline mmdet/models/roi_heads/roi_extractors/generic_roi_extractor.py:73]: `self.pre_module` is a ConvModule, why is there another relu layer?

File: configs/groie/mask_rcnn_r101_fpn_syncbn-backbone_r4_gcb_c3-c5_groie_1x_coco.py [hunk 10]
```diff
@@ -3,7 +3,7 @@
 model = dict(
     roi_head=dict(
         bbox_roi_extractor=dict(
-            type='SumGenericRoiExtractor',
+            type='GenericRoIExtractor',
```
Comment (hellock) [inline configs/groie/mask_rcnn_r101_fpn_syncbn-backbone_r4_gcb_c3-c5_groie_1x_coco.py:6]: You may also add the argument `aggregation`.

**Pull request merged**